File: scripts/pipeline/merge_smart_config_kit.py
"""
Merge Smart-Config-Kit supplemental rules into Script Hub rulesets.

Source rules live at:
  rulesets/Sources/custom/SmartConfigKit/shunt-rules/  (Passwall2 shunt-rules format)
  rulesets/Sources/custom/SmartConfigKit/UK.list        (Surge DOMAIN/KEYWORD format)
  rulesets/Sources/custom/SmartConfigKit/CiciAi.list   (Surge DOMAIN/KEYWORD format)

The former fork dependency (Smart-Config-Kit/) has been removed; all source
files are now vendored under rulesets/Sources/custom/SmartConfigKit/.
"""

import os
import re
from collections import defaultdict
from hub.common import Logger, get_project_root, write_file

ROOT = get_project_root()
SCK_DIR = os.path.join(ROOT, "rulesets/Sources/custom/SmartConfigKit/shunt-rules")
CUSTOM_DIR = os.path.join(ROOT, "rulesets/Sources/custom/SmartConfigKit")
SOURCES_DIR = os.path.join(ROOT, "rulesets/Sources/Links")

# Surge-format extra lists vendored inside custom/SmartConfigKit/
EXTRA_LISTS = {
    "UK.list": "StreamEU",
    "CiciAi.list": "AI",
}

# Mapping: SCK shunt-rules file → Script Hub target ruleset
MAPPING = {
    "01-ai-service.list": "AI",
    "02-crypto.list": "Binance",
    "03-payments.list": "PayPal",
    "04-email.list": "SocialMedia",
    "05-im.list": "SocialMedia",
    "06-social.list": "SocialMedia",
    "07-work.list": "GitHub",
    "08-cn-media.list": "Bilibili",
    "09-sea-media.list": "StreamHK",
    "10-us-media.list": "StreamUS",
    "11-hk-media.list": "StreamHK",
    "12-tw-media.list": "StreamTW",
    "13-jp-media.list": "StreamJP",
    "14-eu-media.list": "StreamEU",
    "15-cn-game.list": "Direct",
    "16-intl-game.list": "Gaming",
    "17-search.list": "Google",
    "18-dev.list": "GitHub",
    "19-microsoft.list": "Microsoft",
    "20-apple.list": "Apple",
    "21-download.list": "Direct",
    "22-cloud-cdn.list": "CDN",
    "24-cn-site.list": "Direct",
    "25-gfw.list": "GlobalProxy",
    "26-intl-site.list": "GlobalProxy",
    "27-final.list": "GlobalProxy",
    "28-ads.list": "AdBlock",
}

_SURGE_RULE = re.compile(
    r"^(DOMAIN(?:-SUFFIX|-KEYWORD|-WILDCARD)?|IP-CIDR6?|PROCESS-NAME),(.+)$",
    re.IGNORECASE,
)


def _parse_shunt(filepath: str) -> list[str]:
    """Parse Passwall2/Xray domain: / full: / keyword: syntax → Surge rules."""
    rules: list[str] = []
    if not os.path.exists(filepath):
        return rules
    with open(filepath, "r", encoding="utf-8") as f:
        for raw in f:
            line = raw.strip()
            if not line or line.startswith("#"):
                continue
            if line.startswith("domain:"):
                rules.append(f"DOMAIN-SUFFIX,{line[7:]}")
            elif line.startswith("full:"):
                rules.append(f"DOMAIN,{line[5:]}")
            elif line.startswith("keyword:"):
                rules.append(f"DOMAIN-KEYWORD,{line[8:]}")
            # geosite:xxx — ignored; script_hub syncs geosite separately via MetaCubeX
    return rules


def _parse_surge_list(filepath: str) -> list[str]:
    """Parse a Surge-format .list file, keeping DOMAIN/IP/PROCESS lines."""
    rules: list[str] = []
    if not os.path.exists(filepath):
        return rules
    with open(filepath, "r", encoding="utf-8") as f:
        for raw in f:
            line = raw.strip()
            if not line or line.startswith("#"):
                continue
            if _SURGE_RULE.match(line):
                rules.append(line)
    return rules


def merge() -> None:
    Logger.section("Merging Smart-Config-Kit supplemental rules")
    os.makedirs(CUSTOM_DIR, exist_ok=True)

    merged: dict[str, dict] = defaultdict(lambda: {"rules": set(), "sources": []})

    # --- shunt-rules (Passwall2 format) ---
    for sck_file, target in MAPPING.items():
        path = os.path.join(SCK_DIR, sck_file)
        rules = _parse_shunt(path)
        if not rules:
            continue
        merged[target]["rules"].update(rules)
        merged[target]["sources"].append(sck_file)
        Logger.success(f"  {len(rules):4d} rules  {sck_file} → {target}")

    # --- extra Surge-format lists ---
    for list_file, target in EXTRA_LISTS.items():
        path = os.path.join(CUSTOM_DIR, list_file)
        rules = _parse_surge_list(path)
        if not rules:
            continue
        merged[target]["rules"].update(rules)
        merged[target]["sources"].append(list_file)
        Logger.success(f"  {len(rules):4d} rules  {list_file} → {target}")

    # --- write SCK_*.txt and link into *_sources.txt ---
    for target, payload in sorted(merged.items()):
        custom_file = f"SCK_{target}.txt"
        custom_path = os.path.join(CUSTOM_DIR, custom_file)

        header = "# Smart-Config-Kit supplement: " + ", ".join(payload["sources"]) + "\n"
        write_file(custom_path, header + "\n".join(sorted(payload["rules"])) + "\n")
        Logger.success(f"  Wrote {len(payload['rules'])} rules → {custom_file}")

        sources_file = os.path.join(SOURCES_DIR, f"{target}_sources.txt")
        if not os.path.exists(sources_file):
            continue
        with open(sources_file, "r", encoding="utf-8") as f:
            content = f.read()
        relative_link = f"../custom/SmartConfigKit/{custom_file}"
        if relative_link not in content:
            with open(sources_file, "a", encoding="utf-8") as f:
                f.write(f"\n# Smart-Config-Kit supplement\n{relative_link}\n")
            Logger.info(f"  Linked {custom_file} → {target}_sources.txt")


if __name__ == "__main__":
    merge()

File: go_scripts/pkg/pipeline/merge_smart_config_kit.go
package pipeline

import (
	"fmt"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/nyamiiko/script_hub/go_scripts/pkg/hub"
)

var (
	sckDir     = hub.SCK_SHUNT_RULES_DIR
	customDir  = hub.SMART_CONFIG_KIT_DIR
	sourcesDir = hub.LINKS_DIR

	extraLists = map[string]string{
		"UK.list":     "StreamEU",
		"CiciAi.list": "AI",
	}

	mapping = map[string]string{
		"01-ai-service.list": "AI",
		"02-crypto.list":     "Binance",
		"03-payments.list":   "PayPal",
		"04-email.list":      "SocialMedia",
		"05-im.list":         "SocialMedia",
		"06-social.list":     "SocialMedia",
		"07-work.list":       "GitHub",
		"08-cn-media.list":   "Bilibili",
		"09-sea-media.list":  "StreamHK",
		"10-us-media.list":   "StreamUS",
		"11-hk-media.list":   "StreamHK",
		"12-tw-media.list":   "StreamTW",
		"13-jp-media.list":   "StreamJP",
		"14-eu-media.list":   "StreamEU",
		"15-cn-game.list":    "Direct",
		"16-intl-game.list":  "Gaming",
		"17-search.list":     "Google",
		"18-dev.list":        "GitHub",
		"19-microsoft.list":  "Microsoft",
		"20-apple.list":      "Apple",
		"21-download.list":   "Direct",
		"22-cloud-cdn.list":  "CDN",
		"24-cn-site.list":    "Direct",
		"25-gfw.list":        "GlobalProxy",
		"26-intl-site.list":  "GlobalProxy",
		"27-final.list":      "GlobalProxy",
		"28-ads.list":        "AdBlock",
	}

	surgeRuleRe = regexp.MustCompile(`(?i)^(DOMAIN(?:-SUFFIX|-KEYWORD|-WILDCARD)?|IP-CIDR6?|PROCESS-NAME),(.+)$`)
)

func parseShunt(filepath string) []string {
	var rules []string
	if !hub.ValidateFileExists(filepath, "") {
		return rules
	}

	content := hub.ReadFileString(filepath)
	lines := strings.Split(content, "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		if strings.HasPrefix(line, "domain:") {
			rules = append(rules, fmt.Sprintf("DOMAIN-SUFFIX,%s", line[7:]))
		} else if strings.HasPrefix(line, "full:") {
			rules = append(rules, fmt.Sprintf("DOMAIN,%s", line[5:]))
		} else if strings.HasPrefix(line, "keyword:") {
			rules = append(rules, fmt.Sprintf("DOMAIN-KEYWORD,%s", line[8:]))
		}
	}
	return rules
}

func parseSurgeList(filepath string) []string {
	var rules []string
	if !hub.ValidateFileExists(filepath, "") {
		return rules
	}

	content := hub.ReadFileString(filepath)
	lines := strings.Split(content, "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		if surgeRuleRe.MatchString(line) {
			rules = append(rules, line)
		}
	}
	return rules
}

func MergeSmartConfigKit() {
	hub.Section("Merging Smart-Config-Kit supplemental rules")
	hub.EnsureDir(customDir)

	type TargetData struct {
		rules   map[string]bool
		sources []string
	}
	merged := make(map[string]*TargetData)

	ensureTarget := func(t string) *TargetData {
		if merged[t] == nil {
			merged[t] = &TargetData{rules: make(map[string]bool)}
		}
		return merged[t]
	}

	for sckFile, target := range mapping {
		path := filepath.Join(sckDir, sckFile)
		rules := parseShunt(path)
		if len(rules) == 0 {
			continue
		}
		td := ensureTarget(target)
		for _, r := range rules {
			td.rules[r] = true
		}
		td.sources = append(td.sources, sckFile)
		hub.Success(fmt.Sprintf("  %4d rules  %s → %s", len(rules), sckFile, target))
	}

	for listFile, target := range extraLists {
		path := filepath.Join(customDir, listFile)
		rules := parseSurgeList(path)
		if len(rules) == 0 {
			continue
		}
		td := ensureTarget(target)
		for _, r := range rules {
			td.rules[r] = true
		}
		td.sources = append(td.sources, listFile)
		hub.Success(fmt.Sprintf("  %4d rules  %s → %s", len(rules), listFile, target))
	}

	var targets []string
	for t := range merged {
		targets = append(targets, t)
	}
	sort.Strings(targets)

	for _, target := range targets {
		payload := merged[target]
		customFile := fmt.Sprintf("SCK_%s.list", target)
		customPath := filepath.Join(customDir, customFile)

		header := fmt.Sprintf("# Smart-Config-Kit supplement: %s\n", strings.Join(payload.sources, ", "))
		var ruleList []string
		for r := range payload.rules {
			ruleList = append(ruleList, r)
		}
		sort.Strings(ruleList)

		content := header + strings.Join(ruleList, "\n") + "\n"
		hub.SafeWriteFile(customPath, content, true)
		hub.Success(fmt.Sprintf("  Wrote %d rules → %s", len(ruleList), customFile))

		sourcesFile := filepath.Join(sourcesDir, fmt.Sprintf("%s_sources.list", target))
		if !hub.ValidateFileExists(sourcesFile, "") {
			continue
		}

		sourcesContent := hub.ReadFileString(sourcesFile)
		relativeLink := fmt.Sprintf("../custom/SmartConfigKit/%s", customFile)
		if !strings.Contains(sourcesContent, relativeLink) {
			sourcesContent += fmt.Sprintf("\n# Smart-Config-Kit supplement\n%s\n", relativeLink)
			hub.SafeWriteFile(sourcesFile, sourcesContent, true)
			hub.Info(fmt.Sprintf("  Linked %s → %s_sources.list", customFile, target))
		}
	}
}
